File: libs/python/edgecommons/messaging/providers/greengrass/ipc_subscription_handler.py
import logging
from typing import Callable
from awsiot.greengrasscoreipc.model import SubscriptionResponseMessage
from edgecommons.messaging.message import Message
from edgecommons.messaging.providers.greengrass.subscription_handler import (
    SubscriptionHandler,
)

logger = logging.getLogger("IpcSubscriptionHandler")


class IpcSubscriptionHandler(SubscriptionHandler):
    def __init__(
        self,
        topic_filter,
        callback: Callable[[str, Message], None],
        max_concurrency: int = None,
        max_messages: int = None,
    ):
        super().__init__(topic_filter, callback, max_concurrency, max_messages)

    def parse_raw_payload(self, event: SubscriptionResponseMessage):
        if event.binary_message is None:
            topic = event.json_message.context.topic
            logger.warning(
                "Received Greengrass JsonMessage on EdgeCommons subscription topic %s; "
                "ignoring non-protobuf payload",
                topic,
            )
            return None
        else:
            topic = event.binary_message.context.topic
            try:
                return topic, Message.from_bytes(event.binary_message.message)
            except ValueError as error:
                logger.warning(
                    "Problem decoding IPC payload into EdgeCommons protobuf Message on topic %s: "
                    "%s. Ignoring message",
                    topic,
                    error,
                )
                return None

File: libs/java/src/main/java/com/mbreissi/edgecommons/messaging/providers/greengrass/IpcSubscriptionHandler.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.edgecommons.messaging.providers.greengrass;


import com.mbreissi.edgecommons.messaging.Message;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import oshi.util.tuples.Pair;
import software.amazon.awssdk.aws.greengrass.model.SubscriptionResponseMessage;

import java.util.function.BiConsumer;

public class IpcSubscriptionHandler extends SubscriptionHandler<SubscriptionResponseMessage>
{
    protected static final Logger LOGGER = LogManager.getLogger(IpcSubscriptionHandler.class);

    public IpcSubscriptionHandler(String topicFilter, BiConsumer<String, Message> callback, int maxConcurrency, int maxMessages)
    {
        super(topicFilter, callback, maxConcurrency, maxMessages);
    }

    @Override
    Pair<String, Message> parseRawPayload(SubscriptionResponseMessage subscriptionResponseMessage)
    {
        LOGGER.debug("IPC Message received on subscription to topic filter '{}'", topicFilter);
        Pair<String, Message> retVal = null;
        try
        {
            String topic;
            if (subscriptionResponseMessage.getJsonMessage() != null)
            {
                topic = subscriptionResponseMessage.getJsonMessage().getContext().getTopic();
                LOGGER.warn("Received Greengrass JsonMessage on EdgeCommons subscription topic {}; ignoring non-protobuf payload",
                        topic);
                return null;
            }
            else
            {
                topic = subscriptionResponseMessage.getBinaryMessage().getContext().getTopic();
                LOGGER.trace("Received binary EdgeCommons message on topic {}", topic);
                retVal = new Pair<>(topic, Message.fromBytes(subscriptionResponseMessage.getBinaryMessage().getMessage()));
            }
        }
        catch (IllegalArgumentException e)
        {
            LOGGER.warn("Problem decoding IPC payload into EdgeCommons protobuf Message on topic {}: {}. Ignoring message",
                    topicFilter, e.toString());
        }
        catch (NullPointerException e)
        {
            LOGGER.error("Null pointer encountered while processing IPC payload on topic {}: {}. Ignoring message",
                    topicFilter, e.toString());
        }
        return retVal;
    }
}
